### ISSUE TITLE
PMD 6.0.0 Readyness

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#709]
The following issues/task I'd like to have resolved, before cutting a release:

- [x] Ruleset Categories - in progress #697 
    - [x] Create `/category/<lang>/categories.properties` (see [comment on 697](https://github.com/pmd/pmd/pull/697#issuecomment-341186233)) -> PR #716 
    - [x] Consider old rulesets as deprecated, include all should still be possible (see [comment on 697](https://github.com/pmd/pmd/pull/697#issuecomment-341815243)) -> PR #731 
    - [x] Update Documentation for new categories, mention categories in Release Notes -> PR #738 
    - [x] Documentation: Should we also generate pages about the "old" rulesets? -> PR #729 

- [ ] Impact on 3rd-party integrations
    - [ ] Eclipse Plugin
    - [ ] maven-pmd-plugin, especially
        - [x] [reporting processing errors](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/PMD.java#L282) / [MPMD-246 Output details of processing errors](https://issues.apache.org/jira/browse/MPMD-246) - finished, depends on PMD release - [see this changes](https://github.com/adangel/maven-plugins/compare/MPMD-247...adangel:MPMD-246)
        - [x] [MPMD-247 Upgrade to PMD 6.0.0](https://issues.apache.org/jira/browse/MPMD-247) - finished, depends on PMD release - [see this changes](https://github.com/adangel/maven-plugins/compare/trunk...adangel:MPMD-247)
    - [x] gradle
    - [ ] Code Climate
    - [ ] Codacy

- [x] Rule Merging/Deprecation
    - [x] java: Use NPathComplexity from metrics to replace existing rule in design -> PR #722
    - [x] java: Use CyclomaticComplexity from metrics to replace existing rule in design -> PR #734 
    - [x] java: Log Guarding Rules #457 -> PR #772 
    - [x] java: Unnecessary(Final)Modifier #411 -> PR #717
    - [x] java: OptimizableToArrayCall (see https://sourceforge.net/p/pmd/bugs/1454/) -> PR #775 
    - [x] java: Emtpy(Static)Initializer  (both are in errorprone now) -> PR #747 
    - [x] java: EmptyCatchBlock: By default ignore "ignored|expected" exception names? -> PR #748 

- [x] Other
    - [x] Fix FUTURES in [Rule](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/Rule.java) - more bean-like getters/setters -> PR #756 
    - [x] Fix UselessParentheses - if not already done. See [AbstractJavaAccessNode.java#123](https://github.com/pmd/pmd/blob/c466179c2ac1293664b7a49df02efe0f065263e3/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractJavaAccessNode.java#L123) - ok, already done
    - [x] NPE in InvalidSlf4jMessageFormat #721 -> PR #726

**Comment by up2go-rsoesemann** [comment]
I can take care about the impact on 3rd party integrations:

1. Codacy is aware of an upcoming 6.0. I also added an issue to their repo to ask about timelines and potential roadbloack regarding new PMD features here: https://github.com/codacy/codacy-pmdjava/issues/43

2. Same with Code Climate. As I am the sole maintainer I will take a bit longer but start integrating it when available. I guess the steps as described in my wiki also apply to 6.0. Or is there something different with the new Jorje?

https://github.com/Up2Go/codeclimate-apexmetrics/wiki/Replace-PMD-used-by-the-engine

It would be great if you could help me get PMD in and adjust the Wiki for furture updates.

**Comment by jsotuyod** [comment]
> - [x] Fix UselessParentheses - if not already done. See AbstractJavaAccessNode.java#123 - ok, already done

I guess the need for #648 is real

**Comment by jsotuyod** [comment]
> - [ ] Documentation: Should we also generate pages about the "old" rulesets?

Probably, but completely different to what they used to be. A separate section to make no mistakes, no rule details, just a "This ruleset is for backwards compatibility for XXX, it includes the following rules".

Deprecation messages can be later included into those, but probably not before we have our new getting started rulesets to help in the transition.

This I believe should also work with the new getting started rulesets:

*"This is a standard ruleset for Android applications. It includes general Java best-practices, style rules conforming to Square's Android style, and Android-specific rules."*

**Comment by up2go-rsoesemann** [comment]
@adangel can you please answer this question by Codacy: https://github.com/codacy/codacy-pmdjava/issues/43#issuecomment-342110023 ?

**Comment by rtfpessoa** [comment]
Can you confirm, that by the release of `6.0.0` all the languages will have the new `/category/<lang>/categories.properties`?

**Comment by up2go-rsoesemann** [comment]
@jsotuyod I guess only thos languages that have more than the CPD, right?

**Comment by adangel** [comment]
@rtfpessoa @up2go-rsoesemann Correct - only those languages, that have actually rules, will have a `/category/<lang>/categories.properties` file.

**Comment by up2go-rsoesemann** [comment]
@adangel Are you waiting on anything from me or @rtfpessoa regarding Code Climate and Codacy readyness?

**Comment by jsotuyod** [comment]
@up2go-rsoesemann both tools would need to make sure to take into account the new Apex Jorje jar as per #773

**Comment by up2go-rsoesemann** [comment]
I will take care about Code Climate when 6.0 is released. But not before. @rtfpessoa can you do it for Codacy?

**Comment by jsotuyod** [comment]
Gradle uses the Ant task, it's working as is. We would have to impact access to new features (such as Java 9 support) after we release so they can include proper tests without a snapshot PMD build.

**Comment by adangel** [comment]
FYI - I plan to release PMD 6.0.0 this Friday. I mainly want to make sure, we don't release something, that doesn't run/is not usable without a code change in other environments (such as codacy, code climate, jenkins, ...). On the other hand, if there is a serious problem, we could release a 6.0.1 (which I want to avoid) or wait for 6.1.0 (which I scheduled for end of January now).

**Comment by up2go-rsoesemann** [comment]
@adangel @jsotuyod can we put two other Apex/Salesforce related tooling providers on our list so we can loop them in in case of new releases and disruptive changes?

I'm thinking about the guys behind the major commercial Salesforce IDEs Scott Well of http://www.illuminatedcloud.com/ and Vladimir Gubanowitch of https://welkinsuite.com/

I will ask both to star/watch this repo so we can @mention them.

**Comment by jsotuyod** [comment]
@up2go-rsoesemann of course we can, the more the merrier.

Have they already built PMD into their tools or are they planning to?

**Comment by up2go-rsoesemann** [comment]
Already built in and getting user issues related to pmd

**Comment by SCWells72** [comment]
Yep, I'm one of the devs and have already integrated PMD (in particular PMD Apex) into my product. I'll definitely start watching this repo so I know when I need to react to changes. Thanks!

**Comment by adangel** [comment]
@SCWells72 Awesome! Feel free, to open issues here if you run into problems.

**Comment by adangel** [comment]
Closing this issue, since 6.0.0 is released now!
We'll open a new issue for any problem that arises while integrating the new 6.0.0 into 3rd party projects.

**Comment by up2go-rsoesemann** [comment]
Although closed let me @mention here the guy from the other major Salesforce IDE which integrated PMD Apex and should be in a close loop on big releases: @bickus from TheWelkinSuite
